### PR TITLE
Fix and implement some nio.file functions

### PR DESCRIPTION
**Pull request by hrj**
Things which this PR fixes:
* Path.toRealPath()
* Files.exists()
* Files.createDirectory(), Files.createDirectories()
* Files.list() (this was failing because of a bug in string conversion)

Note that the implementation of `access0` is not complete. It needs to consider the type-of-access.

Also, file creation with NIO API doesn't work yet, because file channels have not been implemented.

For #403.

### CODE REVIEW

**Comment by hrj** [comment]
The test in Firefox is failing because of a "not supported error". I am guessing this is because browserfs doesn't implement:
* `fs.realpath()` and / or
* `fs.access()`

I am not able to run the karma tests locally to confirm / debug further. (karma + firefox are hanging)

**Comment by jvilk** [comment]
It doesn't implement `fs.access`, as I only implemented Node v0.10 `fs` functions, and didn't notice the new ones that have recently snuck in.

`fs.realpath` should work.

**Comment by jvilk** [comment]
Actually, I guess it is in there, just not implemented:

https://github.com/jvilk/BrowserFS/blob/master/src/core/FS.ts#L1426

**Comment by jvilk** [comment]
Did the karma update fix the freezing, btw?

**Comment by hrj** [comment]
Hmm.. in java_io.ts the fs.stat() function seems to be used. Do you recommend using fs.stat() in java_nio.ts as well? Or better to wait for browserfs to implement fs.access() ?

**Comment by hrj** [comment]
> Did the karma update fix the freezing, btw?

nope.. it didn't :(

Not related to this PR, but I pushed it here anyway thinking it's small enough.

**Comment by jvilk** [comment]
I'd accept using `fs.stat` for now. I logged an issue on BFS to implement `access` at some point.

You can also check if you're in the browser or not using `util.are_in_browser()` (which I need to camelcase someday...), and use `fs.access` or `fs.stat`.

**Comment by hrj** [comment]
I updated the PR to use `fs.stat` when running in browser, and use `fs.access` otherwise.

However, it is still failing in the browser test. A directory is showing as non-existing with `lstat`although it does exist. Could the failure be in browser-fs?

**Comment by jvilk** [comment]
It could be a bug in BrowserFS. I'll try to take a look tonight.

**Comment by jvilk** [comment]
Yup, you found a BFS bug.

```
> bfs_fs.mkdirSync('/tmp')
undefined
> bfs_fs.mkdirSync('/tmp/tmpDir')
undefined
> bfs_fs.mkdirSync('/tmp/tmpDir/tmpDir2')
undefined
> bfs_fs.rmdirSync('/tmp/tmpDir')
undefined
> bfs_fs.readdirSync('/tmp')
[]
```

I'll look into it.

**Comment by jvilk** [comment]
@hrj I just pushed a fix to BFS, but will wait until I merge your pending PR to update the NPM module.

**Comment by hrj** [comment]
@jvilk Thanks for the fixes and merges in BFS. Awaiting the BFS revision in NPM. I can then update `package.json` in this PR.

**Comment by jvilk** [comment]
@hrj I'm looking into the unit test bug that you're experiencing right now, as I've reproduced it myself.

https://github.com/jvilk/BrowserFS/issues/132

**Comment by jvilk** [comment]
@hrj Released! Update our `package.json` to the newest version, commit, and then we'll see what Travis says.

**Comment by hrj** [comment]
green at last!

**Comment by hrj** [comment]
Friendly ping: can this be merged now?

**Comment by jvilk** [comment]
@hrj Can you remove any logic from the test that doesn't use `java.nio`? I could be misreading, but it looks like you have active test portions that are simply copied from the `FileOps` test directly, with no changes to make them use `java.nio` (e.g. the first part that checks for a nonexistent path).

The same goes for the commented out code.

Thanks.

File: classes/test/NioFilesPaths.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+  private static FileSystem dfs = FileSystems.getDefault();
+
+  public static void main(String[] args) throws IOException {
+    final String testDir = "./classes/test/data/FileOps";
+    final Path testDirPath = Paths.get("./classes/test/data/FileOps");
+    // I like scopes.
+    {
+      // This file does not exist.
+      final File f = Paths.get("/dfsd/dsfds").toFile();
```
Comment (hrj) [inline classes/test/NioFilesPaths.java]: For this test, the `Path.toFile()` is the important bit. To check if the resulting `File` exists we have to use java.io method, can't use nio there.

File: classes/test/NioFilesPaths.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      System.out.println("Creating temp_readonly.txt: " + f.createNewFile());
+      // BrowserFS doesn't support chmod in its temporary file system anymore.
+      // System.out.println("Marking as read only: " + f.setReadOnly());
+      System.out.println("Can I write to the file?: " + f.canWrite());
+      System.out.println("Can I read the file?: " + f.canRead());
+      // make sure we can open a read-only file with RandomAccessFile
+      RandomAccessFile raf = new RandomAccessFile(f, "r");
+      System.out.println("Deleting file: " + f.delete());
+    }
+    */
```
Comment (hrj) [inline classes/test/NioFilesPaths.java]: I left the above commented tests as TODOs. I should have added a TODO comment there; will do so.

Else, if you want me to simply remove them, I can do that.

Comment (jvilk) [inline classes/test/NioFilesPaths.java]: Gotcha. We can leave them in.

**Pull request merged**